File: algorithm/74. Search a 2D Matrix/74.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int n = matrix.size();
        if(n==0)
            return 0;
        int m = matrix[0].size();
        if (m==0)
            return 0;
        int l = 0, r = m * n - 1;
        while (l != r){
            int mid = (l + r - 1) >> 1;
            if (matrix[mid / m][mid % m] < target)
                l = mid + 1;
            else 
                r = mid;
        }
        return matrix[r / m][r % m] == target;
    }
};

File: algorithm/74. Search a 2D Matrix/74.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix.length;
        if(n == 0)  return false;
        int m = matrix[0].length;
        if(m == 0)  return false;
        int l = 0; int r = m*n-1;
        while(l < r)
        {
            int mid = (l + r)/2;
            if(matrix[mid/m][mid%m] < target)
                l = mid + 1;
            else
                r = mid;
        }
        return matrix[l/m][l%m] == target;
    }
}
